// gt-next/dist/server-dir/buildtime/getTranslationFunction.js
"use strict";
// getTranslationFunction.ts (refactored)
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGT = getGT;
exports.useGT = useGT;
exports.getMessages = getMessages;
exports.useMessages = useMessages;
var getI18NConfig_1 = __importDefault(require("../../config-dir/getI18NConfig"));
var server_1 = require("../../server");
var id_1 = require("generaltranslation/id");
var createErrors_1 = require("../../errors/createErrors");
var internal_1 = require("gt-react/internal");
var internal_2 = require("generaltranslation/internal");
var use_1 = __importDefault(require("../../utils/use"));
function createTranslator(_messages) {
    return __awaiter(this, void 0, void 0, function () {
        // --------- HELPERS --------- //
        /**
         * @description Format message and fallback:
         * (1) format message
         * (2) format fallback
         * (3)
         *   - PRODUCTION: return fallback (unformatted)
         *   - DEVELOPMENT: throw error
         */
        function renderMessageHelper(_a) {
            var _b;
            var message = _a.message, variables = _a.variables, locales = _a.locales, fallback = _a.fallback, id = _a.id, maxChars = _a.maxChars;
            try {
                // (1) Try to format message
                var declaredVars = (0, internal_2.extractVars)(fallback || '');
                var formattedMessage = gtClass.formatMessage(Object.keys(declaredVars).length ? (0, internal_2.condenseVars)(message) : message, {
                    locales: locales,
                    variables: __assign(__assign(__assign({}, variables), declaredVars), (_b = {}, _b[internal_2.VAR_IDENTIFIER] = 'other', _b)),
                });
                var cutoffMessage = gtClass.formatCutoff(formattedMessage, {
                    locales: locales,
                    maxChars: maxChars,
                });
                return cutoffMessage;
            }
            catch (error) {
                if (process.env.NODE_ENV === 'production') {
                    console.warn((0, createErrors_1.createStringRenderWarning)(message, id), 'Error: ', error);
                }
                else {
                    // (3) If no fallback, throw error (non-prod)
                    if (!fallback)
                        throw new Error("".concat((0, createErrors_1.createStringRenderError)(message, id), " Error: ").concat(error));
                    console.error((0, createErrors_1.createStringRenderError)(message, id), 'Error: ', error);
                }
                // (2) If format fails, format fallback
                if (fallback) {
                    return renderMessageHelper({
                        message: fallback,
                        locales: locales,
                        variables: variables,
                        id: id,
                        maxChars: maxChars,
                    });
                }
                // (3) Fallback to original message (unformatted)
                var cutoffMessage = gtClass.formatCutoff(message, {
                    locales: locales,
                    maxChars: maxChars,
                });
                return cutoffMessage; // fallback to original message (unformatted)
            }
        }
        function initializeGT(message, options) {
            if (options === void 0) { options = {}; }
            if (!message || typeof message !== 'string')
                return null;
            var id = options.$id, context = options.$context, maxChars = options.$maxChars, _hash = options.$_hash, variables = __rest(options, ["$id", "$context", "$maxChars", "$_hash"]);
            var renderMessage = function (msg, locales, fallback) {
                return renderMessageHelper({
                    message: msg,
                    locales: locales,
                    variables: variables,
                    id: id,
                    fallback: fallback,
                    maxChars: maxChars,
                });
            };
            var calculateHash = function () {
                return (0, id_1.hashSource)(__assign(__assign(__assign(__assign({ source: (0, internal_2.indexVars)(message) }, (context && { context: context })), (maxChars != null && { maxChars: Math.abs(maxChars) })), (id && { id: id })), { dataFormat: 'ICU' }));
            };
            return {
                id: id,
                context: context,
                maxChars: maxChars,
                _hash: _hash,
                variables: variables,
                calculateHash: calculateHash,
                renderMessage: renderMessage,
            };
        }
        function getTranslationData(calculateHash, id, _hash) {
            var translationEntry;
            var hash = ''; // lazily computed if needed
            if (id)
                translationEntry = translations === null || translations === void 0 ? void 0 : translations[id];
            if (!translationEntry && _hash && (translations === null || translations === void 0 ? void 0 : translations[_hash]) !== undefined) {
                hash = _hash;
                translationEntry = translations === null || translations === void 0 ? void 0 : translations[_hash];
            }
            if (!translationEntry) {
                hash = calculateHash();
                translationEntry = translations === null || translations === void 0 ? void 0 : translations[hash];
            }
            return { translationEntry: translationEntry, hash: hash };
        }
        function scheduleTranslateOnDemand(args) {
            var source = args.source, context = args.context, maxChars = args.maxChars, id = args.id, hash = args.hash, renderMessage = args.renderMessage;
            try {
                I18NConfig.translateIcu({
                    source: (0, internal_2.indexVars)(source),
                    targetLocale: locale,
                    options: __assign(__assign(__assign(__assign({}, (context && { context: context })), (maxChars && { maxChars: maxChars })), (id && { id: id })), { hash: hash }),
                }).then(function (result) {
                    // eslint-disable-next-line no-console
                    console.warn((0, createErrors_1.createTranslationLoadingWarning)(__assign(__assign({}, (id && { id: id })), { source: renderMessage(source, [defaultLocale]), translation: renderMessage(result, [locale, defaultLocale], source) })));
                });
            }
            catch (error) {
                console.warn(error);
            }
        }
        var I18NConfig, locale, defaultLocale, translationRequired, gtClass, translations, _a, preloadedTranslations, preload, gt, m;
        var _this = this;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    I18NConfig = (0, getI18NConfig_1.default)();
                    return [4 /*yield*/, (0, server_1.getLocale)()];
                case 1:
                    locale = _b.sent();
                    defaultLocale = I18NConfig.getDefaultLocale();
                    translationRequired = I18NConfig.requiresTranslation(locale)[0];
                    gtClass = I18NConfig.getGTClass();
                    if (!translationRequired) return [3 /*break*/, 3];
                    return [4 /*yield*/, I18NConfig.getCachedTranslations(locale)];
                case 2:
                    _a = _b.sent();
                    return [3 /*break*/, 4];
                case 3:
                    _a = undefined;
                    _b.label = 4;
                case 4:
                    translations = _a;
                    if (!(internal_1.reactHasUse &&
                        _messages &&
                        I18NConfig.isDevelopmentApiEnabled() &&
                        translationRequired)) return [3 /*break*/, 6];
                    preloadedTranslations = {};
                    preload = function (_a) { return __awaiter(_this, void 0, void 0, function () {
                        var init, id, context, maxChars, _hash, calculateHash, _b, translationEntry, hash, _c, _d, error_1;
                        var message = _a.message, options = __rest(_a, ["message"]);
                        return __generator(this, function (_e) {
                            switch (_e.label) {
                                case 0:
                                    if (!message)
                                        return [2 /*return*/];
                                    init = initializeGT(message, options);
                                    if (!init)
                                        return [2 /*return*/];
                                    id = init.id, context = init.context, maxChars = init.maxChars, _hash = init._hash, calculateHash = init.calculateHash;
                                    _b = getTranslationData(calculateHash, id, _hash), translationEntry = _b.translationEntry, hash = _b.hash;
                                    if (translationEntry)
                                        return [2 /*return*/]; // exists already
                                    _e.label = 1;
                                case 1:
                                    _e.trys.push([1, 3, , 4]);
                                    _c = preloadedTranslations;
                                    _d = hash;
                                    return [4 /*yield*/, I18NConfig.translateIcu({
                                            source: (0, internal_2.indexVars)(message),
                                            targetLocale: locale,
                                            options: __assign(__assign(__assign(__assign({}, (context && { context: context })), (maxChars && { maxChars: maxChars })), (id && { id: id })), { hash: hash }),
                                        })];
                                case 2:
                                    _c[_d] = (_e.sent());
                                    return [3 /*break*/, 4];
                                case 3:
                                    error_1 = _e.sent();
                                    console.warn(error_1);
                                    return [3 /*break*/, 4];
                                case 4: return [2 /*return*/];
                            }
                        });
                    }); };
                    return [4 /*yield*/, Promise.all(_messages.map(preload))];
                case 5:
                    _b.sent();
                    _b.label = 6;
                case 6:
                    gt = function (message, options) {
                        if (options === void 0) { options = {}; }
                        var init = initializeGT(message, options);
                        if (!init)
                            return '';
                        var id = init.id, context = init.context, maxChars = init.maxChars, _hash = init._hash, calculateHash = init.calculateHash, renderMessage = init.renderMessage;
                        // Early: no translation needed
                        if (!translationRequired)
                            return renderMessage(message, [defaultLocale]);
                        var _a = getTranslationData(calculateHash, id, _hash), translationEntry = _a.translationEntry, hash = _a.hash;
                        if (translationEntry) {
                            return renderMessage(translationEntry, [locale, defaultLocale], message);
                        }
                        if (!I18NConfig.isDevelopmentApiEnabled()) {
                            console.warn((0, createErrors_1.createStringTranslationError)(message, id, 't'));
                            return renderMessage(message, [defaultLocale]);
                        }
                        if (!translationEntry && (preloadedTranslations === null || preloadedTranslations === void 0 ? void 0 : preloadedTranslations[hash])) {
                            return renderMessage(preloadedTranslations[hash], [locale, defaultLocale], message);
                        }
                        // On-demand translate
                        scheduleTranslateOnDemand({
                            source: message,
                            context: context,
                            maxChars: maxChars,
                            id: id,
                            hash: hash,
                            renderMessage: renderMessage,
                        });
                        return renderMessage(message, [defaultLocale]);
                    };
                    m = function (encodedMsg, options) {
                        if (!encodedMsg)
                            return encodedMsg;
                        // Try to decode first
                        var decodedOptions = (0, internal_1.decodeOptions)(encodedMsg);
                        // Fallback to t() if not an encoded message
                        if (!decodedOptions || !decodedOptions.$_hash || !decodedOptions.$_source) {
                            return gt(encodedMsg, options);
                        }
                        var $_hash = decodedOptions.$_hash, $_source = decodedOptions.$_source, $context = decodedOptions.$context, $id = decodedOptions.$id, $maxChars = decodedOptions.$maxChars, decodedVariables = __rest(decodedOptions, ["$_hash", "$_source", "$context", "$id", "$maxChars"]);
                        var renderMessage = function (msg, locales, fallback) {
                            return renderMessageHelper({
                                message: msg,
                                locales: locales,
                                variables: decodedVariables,
                                fallback: fallback,
                                maxChars: $maxChars,
                            });
                        };
                        // Early: default locale only
                        if (!translationRequired)
                            return renderMessage($_source, [defaultLocale]);
                        // Translation exists?
                        var translationEntry = translations === null || translations === void 0 ? void 0 : translations[$_hash];
                        if (translationEntry === null) {
                            return renderMessage($_source, [defaultLocale]);
                        }
                        if (translationEntry) {
                            return renderMessage(translationEntry, [locale, defaultLocale], $_source);
                        }
                        // Dev-only paths for loading or preloaded
                        if (!I18NConfig.isDevelopmentApiEnabled()) {
                            console.warn((0, createErrors_1.createStringTranslationError)($_source, (0, internal_1.decodeMsg)(encodedMsg), 'm'));
                            return renderMessage($_source, [defaultLocale]);
                        }
                        if (typeof (preloadedTranslations === null || preloadedTranslations === void 0 ? void 0 : preloadedTranslations[$_hash]) !== 'undefined') {
                            if (preloadedTranslations === null || preloadedTranslations === void 0 ? void 0 : preloadedTranslations[$_hash]) {
                                return renderMessage(preloadedTranslations[$_hash], [locale, defaultLocale], $_source);
                            }
                            return renderMessage($_source, [defaultLocale]);
                        }
                        // On-demand translate
                        scheduleTranslateOnDemand({
                            source: $_source,
                            context: $context,
                            maxChars: $maxChars,
                            id: $id,
                            hash: $_hash,
                            renderMessage: renderMessage,
                        });
                        // Default: return source while translation loads
                        return renderMessage($_source, [defaultLocale]);
                    };
                    return [2 /*return*/, { gt: gt, m: m }];
            }
        });
    });
}
// ---------------- Public API (kept stable) ---------------- //
/**
 * getGT() returns a function that translates an ICU message string.
 *
 * @returns A promise of the t() function used for translating strings
 *
 * @example
 * const t = await getGT();
 * console.log(t('Hello, world!')); // Translates 'Hello, world!'
 */
function getGT(_messages) {
    return __awaiter(this, void 0, void 0, function () {
        var gt;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, createTranslator(_messages)];
                case 1:
                    gt = (_a.sent()).gt;
                    return [2 /*return*/, gt];
            }
        });
    });
}
/**
 * Hook wrapper for getGT
 */
function useGT(_messages) {
    return (0, use_1.default)(getGT(_messages));
}
/**
 * getMessages() returns a function that translates an encoded ICU message string.
 *
 * @returns A promise of the m() function used for translating encoded ICU message strings
 *
 * @example
 * const encodedMsg = msg('Hello, world!')
 * const m = await getMessages();
 * console.log(m(encodedMsg)); // Translates 'Hello, world!'
 */
function getMessages(_messages) {
    return __awaiter(this, void 0, void 0, function () {
        var m;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, createTranslator(_messages)];
                case 1:
                    m = (_a.sent()).m;
                    return [2 /*return*/, m];
            }
        });
    });
}
/**
 * Hook wrapper for getMessages
 */
function useMessages(_messages) {
    return (0, use_1.default)(getMessages(_messages));
}
