// gt-next/dist/config.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initGT = void 0;
exports.withGTConfig = withGTConfig;
var path_1 = __importDefault(require("path"));
var fs_1 = __importDefault(require("fs"));
var defaultWithGTConfigProps_1 = __importStar(require("./config-dir/props/defaultWithGTConfigProps"));
var errors_1 = require("./errors");
var generaltranslation_1 = require("generaltranslation");
var getStableNextVersionInfo_1 = require("./plugin/getStableNextVersionInfo");
var validateCompiler_1 = require("./config-dir/utils/validateCompiler");
var resolveRequestFunctionPaths_1 = require("./config-dir/utils/resolveRequestFunctionPaths");
var resolveConfigFilepath_1 = require("./config-dir/utils/resolveConfigFilepath");
var ssgChecks_1 = require("./plugin/checks/ssgChecks");
var cacheComponentsChecks_1 = require("./plugin/checks/cacheComponentsChecks");
/**
 * Initializes General Translation settings for a Next.js application.
 *
 * Use it in `next.config.js` to enable GT translation functionality as a plugin.
 *
 * @example
 * // In next.config.mjs
 * import { withGTConfig } from 'gt-next/config';
 *
 * export default withGTConfig(nextConfig, {
 *   projectId: 'abc-123',
 *   locales: ['en', 'es', 'fr'],
 *   defaultLocale: 'en'
 * })
 *
 * @param {string|undefined} config - Optional config filepath (defaults to './gt.config.json'). If a file is found, it will be parsed for GT config variables.
 * @param {string|undefined} dictionary - Optional dictionary configuration file path. If a string is provided, it will be used as a path.
 * @param {string} [apiKey=defaultInitGTProps.apiKey] - API key for the GeneralTranslation service. Required if using the default GT base URL.
 * @param {string} [devApiKey=defaultInitGTProps.devApiKey] - API key for dev environment only.
 * @param {string} [projectId=defaultInitGTProps.projectId] - Project ID for the GeneralTranslation service. Required for most functionality.
 * @param {string|null} [runtimeUrl=defaultInitGTProps.runtimeUrl] - The base URL for the GT API. Set to an empty string to disable automatic translations. Set to null to disable.
 * @param {string|null} [cacheUrl=defaultInitGTProps.cacheUrl] - The URL for cached translations. Set to null to disable.
 * @param {string[]|undefined} - Whether to use local translations.
 * @param {string[]} [locales=defaultInitGTProps.locales] - List of supported locales for the application.
 * @param {string} [defaultLocale=defaultInitGTProps.defaultLocale] - The default locale to use if none is specified.
 * @param {string|undefined} [getLocalePath="getLocale"] - The path to the custom getLocale function.
 * @param {string|undefined} [getRegionPath="getRegion"] - The path to the custom getRegion function.
 * @param {string|undefined} [getDomainPath="getDomain"] - The path to the custom getDomain function.
 * @param {object} [renderSettings=defaultInitGTProps.renderSettings] - Render settings for how translations should be handled.
 * @param {number} [cacheExpiryTime] - The time in milliseconds for how long translations should be cached.
 * @param {number} [maxConcurrentRequests=defaultInitGTProps.maxConcurrentRequests] - Maximum number of concurrent requests allowed.
 * @param {number} [maxBatchSize=defaultInitGTProps.maxBatchSize] - Maximum translation requests in the same batch.
 * @param {number} [batchInterval=defaultInitGTProps.batchInterval] - The interval in milliseconds between batched translation requests.
 * @param {boolean} [ignoreBrowserLocales=defaultWithGTConfigProps.ignoreBrowserLocales] - Whether to ignore browser's preferred locales.
 * @param {object} headersAndCookies - Additional headers and cookies that can be passed for extended configuration.
 * @param {boolean} [experimentalEnableSSG=false] - Whether to enable SSG.
 * @param {boolean} [disableSSGWarnings=defaultWithGTConfigProps.disableSSGWarnings] - Whether to disable SSG warnings. (deprecated)
 * @param {string|undefined} [getStaticLocalePath="getStaticLocale"] - The path to the static getLocale function. (deprecated)
 * @param {string|undefined} [getStaticRegionPath="getStaticRegion"] - The path to the static getRegion function. (deprecated)
 * @param {string|undefined} [getStaticDomainPath="getStaticDomain"] - The path to the static getDomain function. (deprecated)
 * @param {boolean} [experimentalLocaleResolution=defaultWithGTConfigProps.experimentalLocaleResolution] - Whether to use special server side locale resolution logic (required for Cached Components).
 * @param {string|undefined} [experimentalLocaleResolutionParam=defaultWithGTConfigProps.experimentalLocaleResolutionParam] - The parameter to use for experimental locale resolution.
 * @param {object} metadata - Additional metadata that can be passed for extended configuration.
 *
 * @param {NextConfig} nextConfig - The Next.js configuration object to extend
 * @param {withGTConfigProps} props - General Translation configuration properties
 * @returns {NextConfig} - An updated Next.js config with GT settings applied
 *
 * @throws {Error} If the project ID is missing and default URLs are used, or if the API key is required and missing.
 */
function withGTConfig(nextConfig, props) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r;
    if (nextConfig === void 0) { nextConfig = {}; }
    if (props === void 0) { props = {}; }
    // ---------- LOAD GT CONFIG FILE ---------- //
    var loadedConfig = {};
    try {
        var configPath = void 0;
        if (props.config) {
            configPath = props.config;
        }
        else if (fs_1.default.existsSync(defaultWithGTConfigProps_1.default.config)) {
            configPath = defaultWithGTConfigProps_1.default.config;
        }
        else if (fs_1.default.existsSync('./.gt/gt.config.json')) {
            // Support config under .gt for parity with .locadex
            configPath = './.gt/gt.config.json';
        }
        else if (fs_1.default.existsSync('./.locadex/gt.config.json')) {
            // Backward compatibility: support legacy .locadex directory
            configPath = './.locadex/gt.config.json';
        }
        if (typeof configPath === 'string' && fs_1.default.existsSync(configPath)) {
            var fileContent = fs_1.default.readFileSync(configPath, 'utf-8');
            loadedConfig = JSON.parse(fileContent);
        }
    }
    catch (error) {
        console.error('Error reading GT config file:', error);
    }
    // ---------- LOAD ENVIRONMENT VARIABLES ---------- //
    // resolve project ID
    var projectId = process.env.GT_PROJECT_ID;
    // resolve API keys
    var envApiKey = process.env.NODE_ENV === 'production'
        ? process.env.GT_API_KEY
        : process.env.GT_DEV_API_KEY || process.env.GT_API_KEY;
    var apiKey, devApiKey;
    if (envApiKey) {
        var apiKeyType = (_a = envApiKey === null || envApiKey === void 0 ? void 0 : envApiKey.split('-')) === null || _a === void 0 ? void 0 : _a[1];
        if (apiKeyType === 'api') {
            apiKey = envApiKey;
        }
        else if (apiKeyType === 'dev') {
            devApiKey = envApiKey;
        }
    }
    // conditionally add environment variables to config
    var envConfig = __assign(__assign(__assign({}, (projectId ? { projectId: projectId } : {})), (apiKey ? { apiKey: apiKey } : {})), (devApiKey ? { devApiKey: devApiKey } : {}));
    // ---------- CHECK FOR CONFIG CONFLICTS ---------- //
    // Check for conflicts between config and params
    var conflicts = Object.entries(loadedConfig)
        .filter(function (_a) {
        var key = _a[0], value = _a[1];
        // Skip if key doesn't exist in props
        if (!(key in props))
            return false;
        var propValue = props[key];
        // Handle null/undefined values
        if (value == null || propValue == null) {
            return value !== propValue;
        }
        // Handle primitive types (string, number, boolean)
        if (typeof value !== 'object') {
            return value !== propValue;
        }
        // Handle arrays (no need for deep equality check)
        if (Array.isArray(value)) {
            if (!Array.isArray(propValue))
                return true;
            if (value.length !== propValue.length)
                return true;
            return value.some(function (v, i) { return v !== propValue[i]; });
        }
        // Handle objects
        if (typeof value === 'object' && typeof propValue === 'object') {
            var valueKeys = Object.keys(value);
            var propKeys = Object.keys(propValue);
            var keys = new Set(__spreadArray(__spreadArray([], valueKeys, true), propKeys, true));
            // Objects must match exactly (no need to go deeper)
            if (valueKeys.length !== propKeys.length)
                return true;
            return !Array.from(keys).every(function (k) { return value[k] === propValue[k]; });
        }
        return false;
    })
        .map(function (_a) {
        var key = _a[0], value = _a[1];
        return "- Key: ".concat(key, " Next Config: ").concat(JSON.stringify(props[key]), " does not match GT Config: ").concat(JSON.stringify(value));
    });
    if (conflicts.length) {
        throw new Error((0, errors_1.conflictingConfigurationBuildError)(conflicts));
    }
    // ---------- MERGE CONFIGS ---------- //
    // Merge cookie and header names
    var mergedHeadersAndCookies = __assign(__assign({}, defaultWithGTConfigProps_1.default.headersAndCookies), props.headersAndCookies);
    // Merge experimentalSwcPluginOptions
    var mergedExperimentalCompilerOptions = __assign(__assign(__assign({}, defaultWithGTConfigProps_1.default.experimentalCompilerOptions), props.experimentalSwcPluginOptions), props.experimentalCompilerOptions);
    // precedence: input > env > config file > defaults
    var mergedConfig = __assign(__assign(__assign(__assign(__assign({}, defaultWithGTConfigProps_1.default), loadedConfig), envConfig), props), { headersAndCookies: mergedHeadersAndCookies, experimentalCompilerOptions: mergedExperimentalCompilerOptions, _usingPlugin: true });
    // clear up any issues with the compiler options
    (0, validateCompiler_1.validateCompiler)(mergedConfig);
    // ----------- RESOLVE ANY EXTERNAL FILES ----------- //
    // Resolve wasm filepath
    var turboPackEnabled = !!process.env.TURBOPACK;
    var resolvedWasmFilePath = '';
    if (((_b = mergedConfig.experimentalCompilerOptions) === null || _b === void 0 ? void 0 : _b.type) === 'swc') {
        try {
            if (turboPackEnabled) {
                var absolutePath = path_1.default.resolve(__dirname, './gt_swc_plugin.wasm');
                resolvedWasmFilePath =
                    './' + path_1.default.relative(process.cwd(), absolutePath).replace(/\\/g, '/');
            }
            else {
                resolvedWasmFilePath = path_1.default.resolve(__dirname, './gt_swc_plugin.wasm');
            }
        }
        catch (error) {
            console.error((0, errors_1.createGTCompilerUnresolvedWarning)('swc'), 'Error message:', error);
            resolvedWasmFilePath = '';
            mergedConfig.experimentalCompilerOptions.type = 'none';
        }
    }
    // Resolve dictionary filepath
    var resolvedDictionaryFilePath = typeof mergedConfig.dictionary === 'string'
        ? mergedConfig.dictionary
        : (0, resolveConfigFilepath_1.resolveConfigFilepath)('dictionary', ['.ts', '.js', '.json']); // fallback to dictionary
    // Check [defaultLocale].json file
    if (!resolvedDictionaryFilePath && mergedConfig.defaultLocale) {
        resolvedDictionaryFilePath = (0, resolveConfigFilepath_1.resolveConfigFilepath)(mergedConfig.defaultLocale, ['.json']);
        // Check [defaultLanguageCode].json file
        if (!resolvedDictionaryFilePath) {
            var defaultLanguage = (_c = (0, generaltranslation_1.getLocaleProperties)(mergedConfig.defaultLocale)) === null || _c === void 0 ? void 0 : _c.languageCode;
            if (defaultLanguage && defaultLanguage !== mergedConfig.defaultLocale) {
                resolvedDictionaryFilePath = (0, resolveConfigFilepath_1.resolveConfigFilepath)(defaultLanguage, [
                    '.json',
                ]);
            }
        }
    }
    // Get the type of dictionary file
    var resolvedDictionaryFilePathType = resolvedDictionaryFilePath
        ? path_1.default.extname(resolvedDictionaryFilePath)
        : undefined;
    if (resolvedDictionaryFilePathType) {
        mergedConfig._dictionaryFileType = resolvedDictionaryFilePathType;
    }
    // Resolve custom dictionary loader path
    var customLoadDictionaryPath = typeof mergedConfig.loadDictionaryPath === 'string'
        ? mergedConfig.loadDictionaryPath
        : (0, resolveConfigFilepath_1.resolveConfigFilepath)('loadDictionary');
    // Resolve custom translation loader path
    var customLoadTranslationsPath = typeof mergedConfig.loadTranslationsPath === 'string'
        ? mergedConfig.loadTranslationsPath
        : (0, resolveConfigFilepath_1.resolveConfigFilepath)('loadTranslations');
    // Resolve request function paths
    var requestFunctionPaths = (0, resolveRequestFunctionPaths_1.resolveRequestFunctionPaths)(mergedConfig);
    // Warn if found in /app directory
    if (!resolvedDictionaryFilePath &&
        (0, resolveConfigFilepath_1.resolveConfigFilepath)('dictionary', ['.ts', '.js', '.json'], undefined, [
            './app',
            './src/app',
        ])) {
        console.warn((0, errors_1.createBadFilepathWarning)('dictionary', ['./app', './src/app']));
    }
    if (!customLoadDictionaryPath &&
        (0, resolveConfigFilepath_1.resolveConfigFilepath)('loadDictionary', ['.ts', '.js', '.json'], undefined, ['./app', './src/app'])) {
        console.warn((0, errors_1.createBadFilepathWarning)('loadDictionary', ['./app', './src/app']));
    }
    if (!customLoadTranslationsPath &&
        (0, resolveConfigFilepath_1.resolveConfigFilepath)('loadTranslations', ['.ts', '.js', '.json'], undefined, ['./app', './src/app'])) {
        console.warn((0, errors_1.createBadFilepathWarning)('loadTranslations', ['./app', './src/app']));
    }
    // ----------- LOCALE STANDARDIZATION ----------- //
    // Check if using Services
    var gtRuntimeTranslationEnabled = !!(mergedConfig.runtimeUrl === defaultWithGTConfigProps_1.default.runtimeUrl &&
        ((process.env.NODE_ENV === 'production' && mergedConfig.apiKey) ||
            (process.env.NODE_ENV === 'development' && mergedConfig.devApiKey)));
    var gtRemoteCacheEnabled = !!(mergedConfig.cacheUrl === defaultWithGTConfigProps_1.default.cacheUrl &&
        mergedConfig.loadTranslationsType === 'remote');
    var gtServicesEnabled = !!((gtRuntimeTranslationEnabled || gtRemoteCacheEnabled) &&
        mergedConfig.projectId);
    // Standardize locales
    if (mergedConfig.locales && mergedConfig.defaultLocale) {
        mergedConfig.locales.unshift(mergedConfig.defaultLocale);
    }
    var updatedLocales = [];
    mergedConfig.locales = Array.from(new Set(mergedConfig.locales)).map(function (locale) {
        var updatedLocale = gtServicesEnabled
            ? (0, generaltranslation_1.standardizeLocale)(locale)
            : locale;
        if (updatedLocale !== locale) {
            updatedLocales.push("".concat(locale, " -> ").concat(updatedLocale));
        }
        return updatedLocale;
    });
    // Standardize canonical locales
    var updatedCanonicalLocales = [];
    if (mergedConfig.customMapping) {
        mergedConfig.customMapping = Object.fromEntries(Object.entries(mergedConfig.customMapping).map(function (_a) {
            var key = _a[0], value = _a[1];
            if (typeof value !== 'object' || !('code' in value)) {
                return [key, value];
            }
            var updatedLocale = gtServicesEnabled
                ? (0, generaltranslation_1.standardizeLocale)(value.code)
                : value.code;
            if (updatedLocale !== value.code) {
                updatedCanonicalLocales.push("".concat(key, " -> ").concat(updatedLocale));
            }
            return [
                key,
                __assign(__assign({}, value), { code: updatedLocale }),
            ];
        }));
    }
    // Run SSG checks
    (0, ssgChecks_1.ssgChecks)(mergedConfig, requestFunctionPaths);
    // Run cache component checks
    (0, cacheComponentsChecks_1.cacheComponentsChecks)({
        mergedConfig: mergedConfig,
        nextConfig: nextConfig,
        requestFunctionPaths: requestFunctionPaths,
        localTranslationsEnabled: !!customLoadTranslationsPath,
        localDictionaryEnabled: !!customLoadDictionaryPath,
    });
    // ---------- DERIVED CONFIG ATTRIBUTES ---------- //
    // Local dictionary flag
    if (customLoadDictionaryPath) {
        // Check: file exists if provided
        if (!fs_1.default.existsSync(path_1.default.resolve(customLoadDictionaryPath))) {
            throw new Error((0, errors_1.unresolvedLoadDictionaryBuildError)(customLoadDictionaryPath));
        }
        else {
            mergedConfig.loadDictionaryEnabled = true;
        }
    }
    else {
        mergedConfig.loadDictionaryEnabled = false;
    }
    // Local translations flag
    if (customLoadTranslationsPath) {
        // Check: file exists if provided
        if (!fs_1.default.existsSync(path_1.default.resolve(customLoadTranslationsPath))) {
            throw new Error((0, errors_1.unresolvedLoadTranslationsBuildError)(customLoadTranslationsPath));
        }
        else {
            mergedConfig.loadTranslationsType = 'custom';
        }
    }
    else {
        mergedConfig.loadTranslationsType = 'remote';
    }
    // Set default cache expiry if and only if no dev key
    if (mergedConfig.loadTranslationsType == 'remote' &&
        !mergedConfig.devApiKey &&
        typeof mergedConfig.cacheExpiryTime === 'undefined') {
        mergedConfig.cacheExpiryTime = defaultWithGTConfigProps_1.defaultCacheExpiryTime;
    }
    // ---------- ERROR CHECKS ---------- //
    // Check: using deprecated localeMapping
    if (props.customMapping) {
        console.warn(errors_1.deprecatedLocaleMappingWarning);
    }
    // Check: invalid locale
    if (!mergedConfig.customMapping && gtServicesEnabled) {
        var invalidLocales_1 = [];
        mergedConfig.locales.forEach(function (locale) {
            if (!(0, generaltranslation_1.isValidLocale)(locale)) {
                invalidLocales_1.push(locale);
            }
        });
        if (invalidLocales_1.length) {
            throw new Error((0, errors_1.invalidLocalesError)(invalidLocales_1));
        }
    }
    // Check: invalid canonical locale
    if (mergedConfig.customMapping && gtServicesEnabled) {
        var invalidCanonicalLocales_1 = [];
        mergedConfig.locales.forEach(function (locale) {
            if (!(0, generaltranslation_1.isValidLocale)(locale, mergedConfig.customMapping)) {
                invalidCanonicalLocales_1.push(locale);
            }
        });
        if (invalidCanonicalLocales_1.length) {
            throw new Error((0, errors_1.invalidCanonicalLocalesError)(invalidCanonicalLocales_1));
        }
    }
    // Check: projectId is not required for remote infrastructure, but warn if missing for dev, nothing for prod
    if ((mergedConfig.cacheUrl || mergedConfig.runtimeUrl) &&
        !mergedConfig.projectId &&
        process.env.NODE_ENV === 'development' &&
        mergedConfig.loadTranslationsType === 'remote' &&
        !mergedConfig.loadDictionaryEnabled // skip warn if using local dictionary
    ) {
        console.warn(errors_1.projectIdMissingWarn);
    }
    // Check: dev API key should not be included in production
    if (process.env.NODE_ENV === 'production' && mergedConfig.devApiKey) {
        throw new Error(errors_1.devApiKeyIncludedInProductionError);
    }
    // Check: An API key is required for runtime translation
    if (mergedConfig.projectId && // must have projectId for this check to matter anyways
        mergedConfig.runtimeUrl &&
        !(mergedConfig.apiKey || mergedConfig.devApiKey) &&
        process.env.NODE_ENV === 'development') {
        console.warn(errors_1.APIKeyMissingWarn);
    }
    // Check: if using GT infrastructure, warn about unsupported locales
    if (gtServicesEnabled) {
        // Warn about standardized locales
        if (updatedLocales.length) {
            console.warn((0, errors_1.standardizedLocalesWarning)(updatedLocales));
        }
        // Warn about standardized canonical locales
        if (updatedCanonicalLocales.length) {
            console.warn((0, errors_1.standardizedCanonicalLocalesWarning)(updatedCanonicalLocales));
        }
    }
    // ---------- STORE CONFIGURATIONS ---------- //
    var I18NConfigParams = JSON.stringify(mergedConfig);
    // eslint-disable-next-line @typescript-eslint/no-unused-vars , no-unused-vars
    var _s = mergedConfig.experimentalCompilerOptions || {}, type = _s.type, compilerOptions = __rest(_s, ["type"]);
    var swcPluginEntry = ((_d = mergedConfig.experimentalCompilerOptions) === null || _d === void 0 ? void 0 : _d.type) === 'swc'
        ? [resolvedWasmFilePath, compilerOptions]
        : null;
    var turboAliases = __assign({ 'gt-next/_dictionary': resolvedDictionaryFilePath || '', 'gt-next/_load-translations': customLoadTranslationsPath || '', 'gt-next/_load-dictionary': customLoadDictionaryPath || '' }, Object.fromEntries(Object.entries(requestFunctionPaths).map(function (_a) {
        var functionName = _a[0], path = _a[1];
        return [
            resolveRequestFunctionPaths_1.REQUEST_FUNCTION_ALIASES[functionName],
            path,
        ];
    })));
    // experimental.turbo is deprecated in next@15.3.0.
    // Check for experimental.turbo. If we write to turbopack field, experimental fields will be ignored.
    // Yet, if there are other resolveAlias fields, we don't want to be ignored either.
    var experimentalTurbopack = !(getStableNextVersionInfo_1.turboConfigStable &&
        (!((_e = nextConfig.experimental) === null || _e === void 0 ? void 0 : _e.turbo) || ((_f = nextConfig.turbopack) === null || _f === void 0 ? void 0 : _f.resolveAlias)));
    var config = __assign(__assign(__assign(__assign({}, nextConfig), { env: __assign(__assign(__assign(__assign({}, nextConfig.env), { _GENERALTRANSLATION_I18N_CONFIG_PARAMS: I18NConfigParams }), (resolvedDictionaryFilePathType && {
            _GENERALTRANSLATION_DICTIONARY_FILE_TYPE: resolvedDictionaryFilePathType,
        })), { _GENERALTRANSLATION_LOCAL_DICTIONARY_ENABLED: mergedConfig.loadDictionaryEnabled.toString(), _GENERALTRANSLATION_LOCAL_TRANSLATION_ENABLED: (mergedConfig.loadTranslationsType === 'custom').toString(), _GENERALTRANSLATION_DEFAULT_LOCALE: (mergedConfig.defaultLocale ||
                defaultWithGTConfigProps_1.default.defaultLocale ||
                '').toString(), _GENERALTRANSLATION_GT_SERVICES_ENABLED: gtServicesEnabled.toString(), _GENERALTRANSLATION_IGNORE_BROWSER_LOCALES: ((_g = mergedConfig.ignoreBrowserLocales) === null || _g === void 0 ? void 0 : _g.toString()) ||
                ((_h = defaultWithGTConfigProps_1.default.ignoreBrowserLocales) === null || _h === void 0 ? void 0 : _h.toString()) ||
                'false', _GENERALTRANSLATION_CUSTOM_GET_LOCALE_ENABLED: requestFunctionPaths.getLocale ? 'true' : 'false', _GENERALTRANSLATION_CUSTOM_GET_REGION_ENABLED: requestFunctionPaths.getRegion ? 'true' : 'false', _GENERALTRANSLATION_CUSTOM_GET_DOMAIN_ENABLED: requestFunctionPaths.getDomain ? 'true' : 'false', _GENERALTRANSLATION_STATIC_GET_LOCALE_ENABLED: requestFunctionPaths.getStaticLocale ? 'true' : 'false', _GENERALTRANSLATION_STATIC_GET_REGION_ENABLED: requestFunctionPaths.getStaticRegion ? 'true' : 'false', _GENERALTRANSLATION_STATIC_GET_DOMAIN_ENABLED: requestFunctionPaths.getStaticDomain ? 'true' : 'false', _GENERALTRANSLATION_DISABLE_SSG_WARNINGS: ((_j = mergedConfig.disableSSGWarnings) === null || _j === void 0 ? void 0 : _j.toString()) || 'false', _GENERALTRANSLATION_ENABLE_SSG: ((_k = mergedConfig.experimentalEnableSSG) === null || _k === void 0 ? void 0 : _k.toString()) || 'false', _GENERALTRANSLATION_EXPERIMENTAL_LOCALE_RESOLUTION: ((_l = mergedConfig.experimentalLocaleResolution) === null || _l === void 0 ? void 0 : _l.toString()) || 'false', _GENERALTRANSLATION_EXPERIMENTAL_LOCALE_RESOLUTION_PARAM: mergedConfig.experimentalLocaleResolutionParam }) }), (turboPackEnabled &&
        !experimentalTurbopack && {
        turbopack: __assign(__assign({}, nextConfig.turbopack), { resolveAlias: __assign(__assign({}, (_m = nextConfig.turbopack) === null || _m === void 0 ? void 0 : _m.resolveAlias), turboAliases) }),
    })), { experimental: __assign(__assign(__assign(__assign({}, nextConfig.experimental), (getStableNextVersionInfo_1.rootParamStability === 'experimental' && {
            rootParams: true,
        })), { swcPlugins: __spreadArray(__spreadArray([], (((_o = nextConfig.experimental) === null || _o === void 0 ? void 0 : _o.swcPlugins) || []), true), (swcPluginEntry ? [swcPluginEntry] : []), true) }), (turboPackEnabled &&
            experimentalTurbopack && {
            turbo: __assign(__assign({}, (_p = nextConfig.experimental) === null || _p === void 0 ? void 0 : _p.turbo), { resolveAlias: __assign(__assign({}, (_r = (_q = nextConfig.experimental) === null || _q === void 0 ? void 0 : _q.turbo) === null || _r === void 0 ? void 0 : _r.resolveAlias), turboAliases) }),
        })), webpack: function webpack() {
            var _a;
            var _b = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                _b[_i] = arguments[_i];
            }
            var webpackConfig = _b[0], options = _b[1];
            // Only apply webpack aliases if we're using webpack (not Turbopack)
            if (!turboPackEnabled) {
                // Try to load GT compiler if available
                if (((_a = mergedConfig.experimentalCompilerOptions) === null || _a === void 0 ? void 0 : _a.type) === 'babel') {
                    try {
                        var gtUnplugin = require('@generaltranslation/compiler').webpack;
                        webpackConfig.plugins.unshift(gtUnplugin(mergedConfig.experimentalCompilerOptions || {}));
                    }
                    catch (e) {
                        mergedConfig.experimentalCompilerOptions.type = 'none';
                        console.warn((0, errors_1.createGTCompilerUnresolvedWarning)('babel'), 'Error message:', e);
                    }
                }
                // Disable cache in dev bc people might move around loadTranslations() and loadDictionary() files
                if (process.env.NODE_ENV === 'development') {
                    webpackConfig.cache = false;
                }
                if (resolvedDictionaryFilePath) {
                    webpackConfig.resolve.alias['gt-next/_dictionary'] = path_1.default.resolve(webpackConfig.context, resolvedDictionaryFilePath);
                }
                if (customLoadTranslationsPath) {
                    webpackConfig.resolve.alias["gt-next/_load-translations"] =
                        path_1.default.resolve(webpackConfig.context, customLoadTranslationsPath);
                }
                if (customLoadDictionaryPath) {
                    webpackConfig.resolve.alias["gt-next/_load-dictionary"] =
                        path_1.default.resolve(webpackConfig.context, customLoadDictionaryPath);
                }
                for (var _c = 0, _d = Object.entries(requestFunctionPaths); _c < _d.length; _c++) {
                    var _e = _d[_c], functionName = _e[0], pathString = _e[1];
                    var key = resolveRequestFunctionPaths_1.REQUEST_FUNCTION_ALIASES[functionName];
                    webpackConfig.resolve.alias[key] = path_1.default.resolve(webpackConfig.context, pathString);
                }
            }
            if (typeof (nextConfig === null || nextConfig === void 0 ? void 0 : nextConfig.webpack) === 'function') {
                return nextConfig.webpack(webpackConfig, options);
            }
            return webpackConfig;
        } });
    return config;
}
// Keep initGT for backward compatibility
var initGT = function (props) { return function (nextConfig) {
    return withGTConfig(nextConfig, props);
}; };
exports.initGT = initGT;
