// gt-next/dist/index.client.js
"use strict";
'use client';
Object.defineProperty(exports, "__esModule", { value: true });
exports.gtFallback = exports.mFallback = exports.decodeVars = exports.declareVar = exports.declareStatic = exports.decodeOptions = exports.decodeMsg = exports.msg = exports.useMessages = exports.useLocaleDirection = exports.useLocaleProperties = exports.useGTClass = exports.useDefaultLocale = exports.useLocales = exports.useLocale = exports.useTranslations = exports.useGT = exports.RegionSelector = exports.LocaleSelector = exports.Plural = exports.Branch = exports.Static = exports.DateTime = exports.Currency = exports.Num = exports.Var = exports.T = void 0;
exports.GTProvider = GTProvider;
exports.Tx = Tx;
var client_1 = require("gt-react/client");
Object.defineProperty(exports, "Var", { enumerable: true, get: function () { return client_1.Var; } });
Object.defineProperty(exports, "Num", { enumerable: true, get: function () { return client_1.Num; } });
Object.defineProperty(exports, "Currency", { enumerable: true, get: function () { return client_1.Currency; } });
Object.defineProperty(exports, "DateTime", { enumerable: true, get: function () { return client_1.DateTime; } });
Object.defineProperty(exports, "Static", { enumerable: true, get: function () { return client_1.Static; } });
Object.defineProperty(exports, "T", { enumerable: true, get: function () { return client_1.T; } });
Object.defineProperty(exports, "Branch", { enumerable: true, get: function () { return client_1.Branch; } });
Object.defineProperty(exports, "Plural", { enumerable: true, get: function () { return client_1.Plural; } });
Object.defineProperty(exports, "LocaleSelector", { enumerable: true, get: function () { return client_1.LocaleSelector; } });
Object.defineProperty(exports, "RegionSelector", { enumerable: true, get: function () { return client_1.RegionSelector; } });
Object.defineProperty(exports, "useGT", { enumerable: true, get: function () { return client_1.useGT; } });
Object.defineProperty(exports, "useTranslations", { enumerable: true, get: function () { return client_1.useTranslations; } });
Object.defineProperty(exports, "useLocale", { enumerable: true, get: function () { return client_1.useLocale; } });
Object.defineProperty(exports, "useLocales", { enumerable: true, get: function () { return client_1.useLocales; } });
Object.defineProperty(exports, "useDefaultLocale", { enumerable: true, get: function () { return client_1.useDefaultLocale; } });
Object.defineProperty(exports, "useGTClass", { enumerable: true, get: function () { return client_1.useGTClass; } });
Object.defineProperty(exports, "useLocaleProperties", { enumerable: true, get: function () { return client_1.useLocaleProperties; } });
Object.defineProperty(exports, "useLocaleDirection", { enumerable: true, get: function () { return client_1.useLocaleDirection; } });
Object.defineProperty(exports, "useMessages", { enumerable: true, get: function () { return client_1.useMessages; } });
Object.defineProperty(exports, "msg", { enumerable: true, get: function () { return client_1.msg; } });
Object.defineProperty(exports, "decodeMsg", { enumerable: true, get: function () { return client_1.decodeMsg; } });
Object.defineProperty(exports, "decodeOptions", { enumerable: true, get: function () { return client_1.decodeOptions; } });
Object.defineProperty(exports, "declareStatic", { enumerable: true, get: function () { return client_1.declareStatic; } });
Object.defineProperty(exports, "declareVar", { enumerable: true, get: function () { return client_1.declareVar; } });
Object.defineProperty(exports, "decodeVars", { enumerable: true, get: function () { return client_1.decodeVars; } });
Object.defineProperty(exports, "mFallback", { enumerable: true, get: function () { return client_1.mFallback; } });
Object.defineProperty(exports, "gtFallback", { enumerable: true, get: function () { return client_1.gtFallback; } });
var createErrors_1 = require("./errors/createErrors");
// Mock <GTProvider> which throws an error
function GTProvider() {
    throw new Error(createErrors_1.gtProviderUseClientError);
}
// Mock <Tx> which throws an error
function Tx() {
    throw new Error(createErrors_1.txUseClientError);
}
