// gt-next/dist/index.server.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.gtFallback = exports.mFallback = exports.decodeVars = exports.declareVar = exports.declareStatic = exports.decodeOptions = exports.decodeMsg = exports.msg = exports.useLocaleDirection = exports.useLocale = exports.useMessages = exports.useTranslations = exports.useGT = exports.Plural = exports.Branch = exports.Static = exports.DateTime = exports.Currency = exports.Num = exports.Var = exports.Tx = exports.T = exports.GTProvider = exports.RegionSelector = exports.LocaleSelector = void 0;
exports.useGTClass = useGTClass;
exports.useLocaleProperties = useLocaleProperties;
exports.useLocales = useLocales;
exports.useDefaultLocale = useDefaultLocale;
require("server-only");
var Var_1 = __importDefault(require("./variables/Var"));
exports.Var = Var_1.default;
var Num_1 = __importDefault(require("./variables/Num"));
exports.Num = Num_1.default;
var Currency_1 = __importDefault(require("./variables/Currency"));
exports.Currency = Currency_1.default;
var DateTime_1 = __importDefault(require("./variables/DateTime"));
exports.DateTime = DateTime_1.default;
var T_1 = __importDefault(require("./server-dir/buildtime/T"));
exports.T = T_1.default;
var Branch_1 = __importDefault(require("./branches/Branch"));
exports.Branch = Branch_1.default;
var Plural_1 = __importDefault(require("./branches/Plural"));
exports.Plural = Plural_1.default;
var GTProvider_1 = __importDefault(require("./provider/GTProvider"));
exports.GTProvider = GTProvider_1.default;
var server_1 = require("./server");
Object.defineProperty(exports, "Tx", { enumerable: true, get: function () { return server_1.Tx; } });
var getTranslations_1 = require("./server-dir/buildtime/getTranslations");
Object.defineProperty(exports, "useTranslations", { enumerable: true, get: function () { return getTranslations_1.useTranslations; } });
var getLocale_1 = require("./request/getLocale");
Object.defineProperty(exports, "useLocale", { enumerable: true, get: function () { return getLocale_1.useLocale; } });
var getLocaleDirection_1 = require("./request/getLocaleDirection");
Object.defineProperty(exports, "useLocaleDirection", { enumerable: true, get: function () { return getLocaleDirection_1.useLocaleDirection; } });
var getI18NConfig_1 = __importDefault(require("./config-dir/getI18NConfig"));
var internal_1 = require("gt-react/internal");
Object.defineProperty(exports, "msg", { enumerable: true, get: function () { return internal_1.msg; } });
Object.defineProperty(exports, "decodeMsg", { enumerable: true, get: function () { return internal_1.decodeMsg; } });
Object.defineProperty(exports, "decodeOptions", { enumerable: true, get: function () { return internal_1.decodeOptions; } });
Object.defineProperty(exports, "declareVar", { enumerable: true, get: function () { return internal_1.declareVar; } });
Object.defineProperty(exports, "decodeVars", { enumerable: true, get: function () { return internal_1.decodeVars; } });
Object.defineProperty(exports, "declareStatic", { enumerable: true, get: function () { return internal_1.declareStatic; } });
Object.defineProperty(exports, "Static", { enumerable: true, get: function () { return internal_1.Static; } });
Object.defineProperty(exports, "mFallback", { enumerable: true, get: function () { return internal_1.mFallback; } });
Object.defineProperty(exports, "gtFallback", { enumerable: true, get: function () { return internal_1.gtFallback; } });
var getTranslationFunction_1 = require("./server-dir/buildtime/getTranslationFunction");
Object.defineProperty(exports, "useMessages", { enumerable: true, get: function () { return getTranslationFunction_1.useMessages; } });
Object.defineProperty(exports, "useGT", { enumerable: true, get: function () { return getTranslationFunction_1.useGT; } });
var index_client_1 = require("./index.client");
Object.defineProperty(exports, "LocaleSelector", { enumerable: true, get: function () { return index_client_1.LocaleSelector; } });
Object.defineProperty(exports, "RegionSelector", { enumerable: true, get: function () { return index_client_1.RegionSelector; } });
function useGTClass() {
    return (0, getI18NConfig_1.default)().getGTClass();
}
function useLocaleProperties(locale) {
    return useGTClass().getLocaleProperties(locale);
}
function useLocales() {
    return (0, getI18NConfig_1.default)().getLocales();
}
function useDefaultLocale() {
    return (0, getI18NConfig_1.default)().getDefaultLocale();
}
