// gt-next/dist/server.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLocaleDirection = exports.getRegion = exports.getLocale = exports.Tx = exports.tx = exports.getMessages = exports.getTranslations = exports.getGT = exports.T = exports.GTProvider = void 0;
exports.getDefaultLocale = getDefaultLocale;
exports.getGTClass = getGTClass;
exports.getLocaleProperties = getLocaleProperties;
exports.getLocales = getLocales;
require("server-only");
var T_1 = __importDefault(require("./server-dir/buildtime/T"));
exports.T = T_1.default;
var tx_1 = __importDefault(require("./server-dir/runtime/tx"));
exports.tx = tx_1.default;
var getLocale_1 = require("./request/getLocale");
Object.defineProperty(exports, "getLocale", { enumerable: true, get: function () { return getLocale_1.getLocale; } });
var getRegion_1 = require("./request/getRegion");
Object.defineProperty(exports, "getRegion", { enumerable: true, get: function () { return getRegion_1.getRegion; } });
var getI18NConfig_1 = __importDefault(require("./config-dir/getI18NConfig"));
var getTranslations_1 = require("./server-dir/buildtime/getTranslations");
Object.defineProperty(exports, "getTranslations", { enumerable: true, get: function () { return getTranslations_1.getTranslations; } });
var GTProvider_1 = __importDefault(require("./provider/GTProvider"));
exports.GTProvider = GTProvider_1.default;
var _Tx_1 = __importDefault(require("./server-dir/runtime/_Tx"));
exports.Tx = _Tx_1.default;
var getLocaleDirection_1 = require("./request/getLocaleDirection");
Object.defineProperty(exports, "getLocaleDirection", { enumerable: true, get: function () { return getLocaleDirection_1.getLocaleDirection; } });
var getTranslationFunction_1 = require("./server-dir/buildtime/getTranslationFunction");
Object.defineProperty(exports, "getMessages", { enumerable: true, get: function () { return getTranslationFunction_1.getMessages; } });
Object.defineProperty(exports, "getGT", { enumerable: true, get: function () { return getTranslationFunction_1.getGT; } });
function getDefaultLocale() {
    return (0, getI18NConfig_1.default)().getDefaultLocale();
}
function getGTClass() {
    return (0, getI18NConfig_1.default)().getGTClass();
}
function getLocaleProperties(locale) {
    return getGTClass().getLocaleProperties(locale);
}
function getLocales() {
    return (0, getI18NConfig_1.default)().getLocales();
}
