// gt-next/dist/plugin/getStableNextVersionInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.babelPluginCompatible = exports.swcPluginCompatible = exports.rootParamStability = exports.turboConfigStable = void 0;
var createErrors_1 = require("../errors/createErrors");
var constants_1 = require("./constants");
/**
 * Get the next version of the package.
 */
function getNextVersion() {
    try {
        var pkg = require('next/package.json');
        return pkg.version;
    }
    catch (error) {
        throw new Error((0, createErrors_1.createUnresolvedNextVersionError)(error));
    }
}
/**
 * Get the react version of the package.
 * I am wary of dynamic imports
 */
function getReactVersion() {
    try {
        var pkg = require('react/package.json');
        return pkg.version;
    }
    catch (error) {
        throw new Error((0, createErrors_1.createUnresolvedReactVersionError)(error));
    }
}
/**
 * Compare two package versions.
 *
 * @param a - The first version.
 * @param b - The second version.
 * @returns True if a is greater than or equal to b, false otherwise.
 */
function comparePackageVersion(a, b) {
    var aParts = a.split('.');
    var bParts = b.split('.');
    for (var i = 0; i < aParts.length; i++) {
        var aPart = Number(aParts[i]) || 0;
        var bPart = Number(bParts[i]) || 0;
        if (aPart > bPart) {
            return true;
        }
        else if (aPart < bPart) {
            return false;
        }
    }
    // If all parts are equal, return true
    return true;
}
/**
 * Starting at version next@15.3.0 experimental field in turbo config was deprecated.
 * Shout out to next-intl: https://github.com/amannn/next-intl/pull/1850
 */
exports.turboConfigStable = comparePackageVersion(getNextVersion(), constants_1.STABLE_TURBO_CONFIG_VERSION);
exports.rootParamStability = (function () {
    var nextVersion = getNextVersion();
    // Check if experimental
    if (comparePackageVersion(nextVersion, constants_1.ROOT_PARAM_STABILITY.experimental)) {
        return 'experimental';
    }
    // Check if unstable
    if (comparePackageVersion(nextVersion, constants_1.ROOT_PARAM_STABILITY.unstable)) {
        return 'unstable';
    }
    // return unsupported
    return 'unsupported';
})();
exports.swcPluginCompatible = comparePackageVersion(getNextVersion(), constants_1.SWC_PLUGIN_SUPPORT);
// disable babel plugin if using react <= 16
exports.babelPluginCompatible = comparePackageVersion(getReactVersion(), constants_1.BABEL_PLUGIN_SUPPORT);
