// gt-next/dist/errors/createErrors.js
"use strict";
// ---- ERRORS ---- //
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStringRenderWarning = exports.disablingCompileTimeHashWarning = exports.createGTCompilerUnavailableWarning = exports.createGTCompilerUnresolvedWarning = exports.deprecatedLocaleMappingWarning = exports.standardizedCanonicalLocalesWarning = exports.standardizedLocalesWarning = exports.dictionaryNotFoundWarning = exports.runtimeTranslationTimeoutWarning = exports.createTranslationLoadingWarning = exports.APIKeyMissingWarn = exports.noInitGTWarn = exports.projectIdMissingWarn = exports.createMismatchingHashWarning = exports.createUnsupportedLocalesWarning = exports.createInvalidDictionaryTranslationEntryWarning = exports.createInvalidDictionaryEntryWarning = exports.createNoEntryFoundWarning = exports.usingDefaultsWarning = exports.createBadFilepathWarning = exports.createInvalidIcuDictionaryEntryWarning = exports.createInvalidIcuDictionaryEntryError = exports.invalidCanonicalLocalesError = exports.invalidLocalesError = exports.createStringRenderError = exports.missingVariablesError = exports.txUseClientError = exports.gtProviderUseClientError = exports.typesFileError = exports.conflictingConfigurationBuildError = exports.unresolvedGetLocaleBuildError = exports.unresolvedLoadTranslationsBuildError = exports.unresolvedLoadDictionaryBuildError = exports.unresolvedCustomLoadTranslationsError = exports.unresolvedCustomLoadDictionaryError = exports.dictionaryDisabledError = exports.createDictionarySubsetError = exports.devApiKeyIncludedInProductionError = exports.createRequiredPrefixError = exports.createDictionaryTranslationError = exports.createStringTranslationError = exports.createUnresolvedReactVersionError = exports.createUnresolvedNextVersionError = exports.customLoadDictionaryWarning = exports.customLoadTranslationsError = exports.remoteTranslationsError = void 0;
var generaltranslation_1 = require("generaltranslation");
var constants_1 = require("../plugin/constants");
exports.remoteTranslationsError = 'gt-next Error: fetching remote translation.';
var customLoadTranslationsError = function (locale) {
    if (locale === void 0) { locale = ''; }
    return "gt-next Error: fetching locally stored translations. If you are using a custom loadTranslations(\"".concat(locale, "\"), make sure it is correctly implemented.");
};
exports.customLoadTranslationsError = customLoadTranslationsError;
var customLoadDictionaryWarning = function (locale) {
    if (locale === void 0) { locale = ''; }
    return "gt-next Warning: fetching locally stored translation dictionary. If you are using a custom loadDictionary(\"".concat(locale, "\"), make sure it is correctly implemented.");
};
exports.customLoadDictionaryWarning = customLoadDictionaryWarning;
var createUnresolvedNextVersionError = function (error) {
    return "gt-next Error: Unable to resolve next version. ".concat(error.message);
};
exports.createUnresolvedNextVersionError = createUnresolvedNextVersionError;
var createUnresolvedReactVersionError = function (error) {
    return "gt-next Error: Unable to resolve react version. ".concat(error.message);
};
exports.createUnresolvedReactVersionError = createUnresolvedReactVersionError;
var createStringTranslationError = function (string, id, functionName) {
    if (functionName === void 0) { functionName = 'tx'; }
    return "gt-next string translation error. ".concat(functionName, "(\"").concat(string, "\")").concat(id ? " with id \"".concat(id, "\"") : '', " could not locate translation.");
};
exports.createStringTranslationError = createStringTranslationError;
var createDictionaryTranslationError = function (id) {
    return "gt-next Error: Dictionary translation entry with id: ".concat(id, " could not be found.");
};
exports.createDictionaryTranslationError = createDictionaryTranslationError;
var createRequiredPrefixError = function (id, requiredPrefix) {
    return "gt-next Error: You are using <GTProvider> with a provided prefix id: \"".concat(requiredPrefix, "\", but one of the children of <GTProvider> has the id \"").concat(id, "\". Change the <GTProvider> id prop or your dictionary structure to proceed.");
};
exports.createRequiredPrefixError = createRequiredPrefixError;
exports.devApiKeyIncludedInProductionError = "gt-next Error: You are attempting a production build using a development API key. Replace this API key with a production API key when you build your app for production.";
var createDictionarySubsetError = function (id, functionName) {
    return "gt-next Error: ".concat(functionName, " with id: \"").concat(id, "\". Invalid dictionary entry detected. Make sure you are navigating to the correct subroute of the dictionary with the ID you provide.");
};
exports.createDictionarySubsetError = createDictionarySubsetError;
exports.dictionaryDisabledError = "gt-next Error: You are trying to use a dictionary, but you have not added the withGTConfig() plugin to your app. You must add withGTConfig() to use dictionaries. For more information, visit generaltranslation.com/docs";
exports.unresolvedCustomLoadDictionaryError = "gt-next Error: loadDictionary() was resolved by plug-in but could not be resolved at run time. This usually means that the file was found, but the loadDictionary() function itself could not be resolved.";
exports.unresolvedCustomLoadTranslationsError = "gt-next Error: loadTranslations() was resolved by plug-in but could not be resolved at run time. This usually means that the file was found, but the loadTranslations() function itself could not be resolved.";
var unresolvedLoadDictionaryBuildError = function (path) {
    return "gt-next Error: File defining loadDictionary() could not be resolved at ".concat(path);
};
exports.unresolvedLoadDictionaryBuildError = unresolvedLoadDictionaryBuildError;
var unresolvedLoadTranslationsBuildError = function (path) {
    return "gt-next Error: File defining loadTranslations() function could not be resolved at ".concat(path);
};
exports.unresolvedLoadTranslationsBuildError = unresolvedLoadTranslationsBuildError;
var unresolvedGetLocaleBuildError = function (path) {
    return "gt-next Error: File defining custom getLocale() function could not be resolved at ".concat(path);
};
exports.unresolvedGetLocaleBuildError = unresolvedGetLocaleBuildError;
var conflictingConfigurationBuildError = function (conflicts) {
    return "gt-next Error: Conflicting configuration".concat(conflicts.length > 1 ? 's' : '', " detected. Please resolve the following conflicts before building your app:\n").concat(conflicts.join('\n'));
};
exports.conflictingConfigurationBuildError = conflictingConfigurationBuildError;
exports.typesFileError = "gt-next Error: There is no scenario in which you should be seeing this error.";
exports.gtProviderUseClientError = "You're attempting to import the Next.js <GTProvider> in a client component. " +
    "Are you sure you want to do this? It's better to import <GTProvider> in a file not marked 'use client' so that it can fetch translations on the server. " +
    "If you really need to put <GTProvider> on the client, import <GTClientProvider> from 'gt-next/client' instead (discouraged when using the Next.js App Router).";
exports.txUseClientError = "You're attempting to use the <Tx> runtime translation component in a client component. " +
    "This is currently unsupported. Please use <T> with variables, " +
    "or make sure <Tx> rendered on the server only. ";
var missingVariablesError = function (variables, message) {
    return "gt-next Error: missing variables: \"".concat(variables.join('", "'), "\" in message: \"").concat(message, "\"");
};
exports.missingVariablesError = missingVariablesError;
var createStringRenderError = function (message, id) {
    return "gt-next Error: error rendering string ".concat(id ? "for id: \"".concat(id, "\"") : '', " original message: \"").concat(message, "\"");
};
exports.createStringRenderError = createStringRenderError;
var invalidLocalesError = function (locales) {
    return "gt-next Error: You are using invalid locale codes in your configuration. " +
        "You must either specify a list of valid locales or use \"customMapping\" to " +
        "specify aliases for the following invalid locales: ".concat(locales.join(', '), ".");
};
exports.invalidLocalesError = invalidLocalesError;
var invalidCanonicalLocalesError = function (locales) {
    return "gt-next Error: You are using invalid canonical locale codes in your configuration: ".concat(locales.join(', '), ".");
};
exports.invalidCanonicalLocalesError = invalidCanonicalLocalesError;
var createInvalidIcuDictionaryEntryError = function (id) {
    return "gt-next Error: Invalid ICU string dictionary entry found for id: \"".concat(id, "\"");
};
exports.createInvalidIcuDictionaryEntryError = createInvalidIcuDictionaryEntryError;
// ---- WARNINGS ---- //
var createInvalidIcuDictionaryEntryWarning = function (id) {
    return "gt-next: Invalid ICU string dictionary entry found for id: \"".concat(id, "\"");
};
exports.createInvalidIcuDictionaryEntryWarning = createInvalidIcuDictionaryEntryWarning;
var createBadFilepathWarning = function (filename, dir) {
    return "gt-next: Found ".concat(filename, " in ").concat(dir.join(' or '), " directory. This is not supported. Please move it to your root directory.");
};
exports.createBadFilepathWarning = createBadFilepathWarning;
exports.usingDefaultsWarning = 'gt-next: Unable to access gt-next configuration. Using defaults.';
var createNoEntryFoundWarning = function (id) {
    return "gt-next: No valid dictionary entry found for id: \"".concat(id, "\"");
};
exports.createNoEntryFoundWarning = createNoEntryFoundWarning;
var createInvalidDictionaryEntryWarning = function (id) {
    return "gt-next: Invalid dictionary entry found for id: \"".concat(id, "\"");
};
exports.createInvalidDictionaryEntryWarning = createInvalidDictionaryEntryWarning;
var createInvalidDictionaryTranslationEntryWarning = function (id) {
    return "gt-next: Invalid dictionary translation entry found for id: \"".concat(id, "\"");
};
exports.createInvalidDictionaryTranslationEntryWarning = createInvalidDictionaryTranslationEntryWarning;
var createUnsupportedLocalesWarning = function (locales) {
    return "gt-next: The following locales are currently unsupported by our service: ".concat(locales
        .map(function (locale) {
        var name = (0, generaltranslation_1.getLocaleProperties)(locale).name;
        return "".concat(locale, " (").concat(name, ")");
    })
        .join(', '));
};
exports.createUnsupportedLocalesWarning = createUnsupportedLocalesWarning;
var createMismatchingHashWarning = function (expectedHash, receivedHash) {
    return "gt-next: Mismatching hashes! Expected hash: ".concat(expectedHash, ", but got hash: ").concat(receivedHash, ". We will still render your translation, but make sure to update to the newest version: generaltranslation.com/docs");
};
exports.createMismatchingHashWarning = createMismatchingHashWarning;
exports.projectIdMissingWarn = "gt-next: Project ID missing! Set projectId as GT_PROJECT_ID in your environment or by passing the projectId parameter to withGTConfig(). Find your project ID: generaltranslation.com/dashboard.";
exports.noInitGTWarn = "gt-next: You are running General Translation without the withGTConfig() plugin. " +
    "This means that you are not translating your app. To activate translation, add the withGTConfig() plugin to your app, " +
    "and set the projectId and apiKey in your environment. " +
    "For more information, visit https://generaltranslation.com/docs/next/tutorials/quickstart";
exports.APIKeyMissingWarn = "gt-next (plugin): A Development API key is required for runtime translation!  " +
    "Find your Development API key: generaltranslation.com/dashboard.  " +
    "(Or, disable this warning message by setting runtimeUrl to an empty string which disables runtime translation.)";
var createTranslationLoadingWarning = function (_a) {
    var source = _a.source, translation = _a.translation, id = _a.id;
    return "[DEV ONLY] Warning: gt-next created translation \"".concat(source, "\" -> \"").concat(translation, "\"") +
        (id ? " for id \"".concat(id, "\"") : '') +
        ". " +
        "In development, hot-reloaded translations may not be be displayed until the page is refreshed. " +
        "In production, translations will be preloaded and there won't be a warning.";
};
exports.createTranslationLoadingWarning = createTranslationLoadingWarning;
exports.runtimeTranslationTimeoutWarning = "gt-next: Runtime translation timed out.";
exports.dictionaryNotFoundWarning = "gt-next: Dictionary not found. Make sure you have added a dictionary to your project (either dictionary.js or [defaultLocale].json), and you have added the withGTConfig() plugin.";
var standardizedLocalesWarning = function (locales) {
    return "gt-next: You are using The following locales were standardized: ".concat(locales.join(', '), ".");
};
exports.standardizedLocalesWarning = standardizedLocalesWarning;
var standardizedCanonicalLocalesWarning = function (locales) {
    return "gt-next: You are using The following canonical locales were standardized: ".concat(locales.join(', '), ".");
};
exports.standardizedCanonicalLocalesWarning = standardizedCanonicalLocalesWarning;
exports.deprecatedLocaleMappingWarning = "gt-next: You are using the deprecated localeMapping configuration. Please move \"customMapping\" to your gt.config.json file.";
var createGTCompilerUnresolvedWarning = function (type) {
    return "gt-next (plugin): The GT ".concat(type, " compiler could not be resolved. Skipping compiler optimizations.");
};
exports.createGTCompilerUnresolvedWarning = createGTCompilerUnresolvedWarning;
var createGTCompilerUnavailableWarning = function (type) {
    return type === 'swc'
        ? "gt-next (plugin): The GT swc compiler is compatible with < next@".concat(constants_1.SWC_PLUGIN_SUPPORT, ". Skipping compiler optimizations.")
        : "gt-next (plugin): The GT babel compiler is compatible with turbopack or < react@".concat(constants_1.BABEL_PLUGIN_SUPPORT, ". Skipping compiler optimizations.");
};
exports.createGTCompilerUnavailableWarning = createGTCompilerUnavailableWarning;
exports.disablingCompileTimeHashWarning = "gt-next (plugin): You are disabling compile time hash. This will disable the compiler optimizations.";
var createStringRenderWarning = function (message, id) {
    return "gt-next: failed to render string ".concat(id ? "for id: \"".concat(id, "\"") : '', " original message: \"").concat(message, "\"");
};
exports.createStringRenderWarning = createStringRenderWarning;
