// gt-next/dist/config-dir/utils/resolveRequestFunctionPaths.js
"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.REQUEST_FUNCTION_ALIASES = void 0;
exports.resolveRequestFunctionPaths = resolveRequestFunctionPaths;
var types_1 = require("../../request/types");
var withGTConfigProps_1 = require("../props/withGTConfigProps");
var resolveConfigFilepath_1 = require("./resolveConfigFilepath");
exports.REQUEST_FUNCTION_ALIASES = {
    getLocale: 'gt-next/internal/_getLocale',
    getRegion: 'gt-next/internal/_getRegion',
    getDomain: 'gt-next/internal/_getDomain',
    getStaticLocale: 'gt-next/internal/static/_getLocale',
    getStaticRegion: 'gt-next/internal/static/_getRegion',
    getStaticDomain: 'gt-next/internal/static/_getDomain',
};
/**
 * Resolves paths for request functions (both regular and static variants)
 * @param mergedConfig Configuration object containing path settings
 * @returns Object mapping function names to their resolved paths and enabled status
 */
function resolveRequestFunctionPaths(mergedConfig) {
    var result = {};
    for (var _i = 0, _a = __spreadArray(__spreadArray([], types_1.REQUEST_FUNCTIONS, true), types_1.STATIC_REQUEST_FUNCTIONS, true); _i < _a.length; _i++) {
        var functionName = _a[_i];
        var configKey = withGTConfigProps_1.REQUEST_FUNCTION_TO_CONFIG_KEY[functionName];
        var path = typeof mergedConfig[configKey] === 'string'
            ? mergedConfig[configKey]
            : (0, resolveConfigFilepath_1.resolveConfigFilepath)(functionName, ['.ts', '.js']);
        if (path) {
            result[functionName] = path;
        }
    }
    return result;
}
