// gt-next/dist/request/utils/getRequestFunction.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRequestFunction = getRequestFunction;
var ssg_1 = require("../../errors/ssg");
var next_internal_1 = require("@generaltranslation/next-internal");
var constants_1 = require("../../utils/constants");
var cacheComponents_1 = require("../../errors/cacheComponents");
var getI18NConfig_1 = __importDefault(require("../../config-dir/getI18NConfig"));
/**
 * Given a function type, return the associated request function
 * @param functionName
 */
function getRequestFunction(functionName) {
    var _this = this;
    if (process.env._GENERALTRANSLATION_EXPERIMENTAL_LOCALE_RESOLUTION === 'true') {
        return handleExperimentalLocaleResolution(functionName);
    }
    var _a = getModule(functionName), moduleError = _a.error, module = _a.module;
    if (moduleError) {
        return function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
            return [2 /*return*/, undefined];
        }); }); };
    }
    // Is using custom getRequest function
    var usingCustom = getUsingCustom(functionName);
    if (usingCustom) {
        // Extract an unknown function
        var _b = extractCustomFunction(module, functionName), extractError = _b.error, value = _b.value;
        if (!extractError) {
            return value;
        }
    }
    // Fallback to default function
    return extractDefaultFunction(module);
}
/* ========== HELPERS ========== */
/**
 * Special handler for when experimentalLocaleResolution is enabled
 */
function handleExperimentalLocaleResolution(functionName) {
    var _this = this;
    // handle getLocale
    if (functionName === 'getLocale') {
        return function () { return __awaiter(_this, void 0, void 0, function () {
            var unverifiedLocale, I18NConfig, gt;
            var _a;
            return __generator(this, function (_b) {
                try {
                    unverifiedLocale = (0, next_internal_1.getRootParam)((_a = process.env
                        ._GENERALTRANSLATION_EXPERIMENTAL_LOCALE_RESOLUTION_PARAM) !== null && _a !== void 0 ? _a : constants_1.defaultExperimentalLocaleResolutionParam);
                    I18NConfig = (0, getI18NConfig_1.default)();
                    gt = I18NConfig.getGTClass();
                    return [2 /*return*/, unverifiedLocale && gt.isValidLocale(unverifiedLocale)
                            ? unverifiedLocale
                            : undefined];
                }
                catch (error) {
                    console.warn(cacheComponents_1.experimentalLocaleResolutionError + error);
                    return [2 /*return*/, undefined];
                }
                return [2 /*return*/];
            });
        }); };
    }
    // disable other request functions
    return function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
        return [2 /*return*/, undefined];
    }); }); };
}
/**
 * Given a function name, returns the module for the function
 * @param functionName
 * @returns failed field is for type safety
 */
function getModule(functionName) {
    try {
        var module_1;
        switch (functionName) {
            case 'getLocale':
                module_1 = require('gt-next/internal/_getLocale');
                break;
            case 'getRegion':
                module_1 = require('gt-next/internal/_getRegion');
                break;
            case 'getDomain':
                module_1 = require('gt-next/internal/_getDomain');
                break;
        }
        return {
            error: false,
            module: module_1,
        };
    }
    catch (error) {
        console.warn((0, ssg_1.createGetRequestFunctionWarning)(functionName) + ' Error: ' + error);
        return {
            error: true,
            module: undefined,
        };
    }
}
/**
 * Returns true if using a custom getHeaders function.
 */
function getUsingCustom(functionName) {
    switch (functionName) {
        case 'getLocale':
            return (process.env._GENERALTRANSLATION_CUSTOM_GET_LOCALE_ENABLED === 'true');
        case 'getRegion':
            return (process.env._GENERALTRANSLATION_CUSTOM_GET_REGION_ENABLED === 'true');
        case 'getDomain':
            return (process.env._GENERALTRANSLATION_CUSTOM_GET_DOMAIN_ENABLED === 'true');
    }
}
/**
 * Given a namespace and a function name, returns the custom request function.
 */
function extractCustomFunction(module, functionName) {
    try {
        var undefinedNamespaceError = "expected a custom ".concat(functionName, " function, but got ").concat(module, ".");
        if (module == null) {
            throw new Error(undefinedNamespaceError);
        }
        if (typeof module === 'function') {
            return {
                error: false,
                value: module,
            };
        }
        else if (typeof module === 'object') {
            if ('default' in module) {
                if (typeof module.default === 'function') {
                    return {
                        error: false,
                        value: module.default,
                    };
                }
                else if (typeof module.default === 'object') {
                    return {
                        error: false,
                        value: extractCustomFunctionHelper(module.default, functionName),
                    };
                }
            }
            else {
                return {
                    error: false,
                    value: extractCustomFunctionHelper(module, functionName),
                };
            }
        }
        throw new Error(undefinedNamespaceError);
    }
    catch (error) {
        console.warn((0, ssg_1.createCustomGetRequestFunctionWarning)(functionName) + ' Error: ' + error);
        return {
            error: true,
            value: undefined,
        };
    }
}
/**
 * Helper function to extract the custom function from the namespace.
 */
var extractCustomFunctionHelper = function (module, functionName) {
    var undefinedNamespaceError = "gt-next Error: expected a custom ".concat(functionName, " function, but got ").concat(module, ".");
    if (module == null) {
        throw new Error(undefinedNamespaceError);
    }
    var result = undefined;
    switch (functionName) {
        case 'getLocale':
            if ('getLocale' in module && typeof module.getLocale === 'function') {
                result = module.getLocale;
            }
            break;
        case 'getRegion':
            if ('getRegion' in module && typeof module.getRegion === 'function') {
                result = module.getRegion;
            }
            break;
        case 'getDomain':
            if ('getDomain' in module && typeof module.getDomain === 'function') {
                result = module.getDomain;
            }
            break;
    }
    if (result == null) {
        throw new Error(undefinedNamespaceError);
    }
    return result;
};
/**
 * Get the default function from the module. Because its default, we know the typing is correct.
 */
function extractDefaultFunction(module) {
    return module.default;
}
